src/main/scala/chisel3/util/experimental/OverprovisionWidths.scala

```diff
@@ -0,0 +1,289 @@
+    * @param smallerWidth width to trim to, expected to be smaller than the width of expr
+    * @param expr a ground typed expression to be trimmed
+    * @return the trimmed expression
+    */
+  def trim(smallerWidth: BigInt, expr: Expression): Expression = {
+      val trimmed = PrimOps.set_primop_type(DoPrim(PrimOps.Bits, Seq(expr), Seq(smallerWidth, 0), UnknownType))
+      expr.tpe match {
+        case _: SIntType => PrimOps.set_primop_type(DoPrim(PrimOps.AsSInt, Seq(trimmed), Nil, UnknownType))
+        case _: UIntType => trimmed
+      }
+  }
+
+  /** Builds a map from instance port to desired overprovisioned width
+    * @param module module to build the instance port map for
+    * @param modulePorts map of (moduleName -> (map of (portName -> overprovisioned width)))
+    * @return Maps (instanceName, portName) -> overprovisioned width
+    */
+  def buildInstancePortMap(module: DefModule,
+                           modulePorts: Map[String, Map[String, Int]]
+                          ): Map[(String, String), Int] = {
+    val instancePorts = mutable.HashMap[(String, String), Int]()
+    def onStmt(s: Statement): Unit = s match {
+      case WDefInstance(_, name, module, _) if modulePorts.contains(module) =>
+        modulePorts(module).foreach { case (portName, over) => instancePorts((name, portName)) = over }
+      case other => other foreach onStmt
+    }
+    module foreach onStmt
+    instancePorts.toMap
+  }
+
+  /** Trims any nested expression which has been overprovisioned
+    * @param ports maps portName to overprovisioned width
+    * @param instancePorts maps (instanceName, portName) to overprovisioned width
+    * @param e expression tree to walk and trim as necessary
+    * @return
+    */
+  def trimExpressions(ports: Map[String, Int], instancePorts: Map[(String, String), Int])(e: Expression): Expression = {
+    (isOverprovisioned(ports, instancePorts)(e), Utils.flow(e), e.tpe) match {
+      case (None, _, _)     => e map trimExpressions(ports, instancePorts)
+      case (_, SinkFlow, _) => e map trimExpressions(ports, instancePorts)
+      case (Some(over), SourceFlow, GroundType(IntWidth(orig))) => if(orig == over) e else trim(orig, e)
+      case other => firrtl.Utils.throwInternalError(s"Unexpected case in match: $other")
+    }
+  }
+
+  /** Trims all expressions in a module which have been overprovisioned
+    * @param module module to trim
+    * @param modulePorts map of (moduleName -> (map of (portName -> overprovisioned width)))
+    * @return module with trimmed expressions
+    */
+  def trimModule(module: DefModule, modulePorts: Map[String, Map[String, Int]]): DefModule = {
+    val instancePorts = buildInstancePortMap(module, modulePorts)
+    val ports = modulePorts.get(module.name)
+    def onStatement(s: Statement): Statement = s map onStatement map trimExpressions(ports.get, instancePorts)
+    if(ports.nonEmpty) module map onStatement else module
+  }
```

Comment:
If we don't want to run `InferTypes`, we could fix the `Expressions` in `trimExpressions` (ie. giving them the correct type). Because they are being trimmed anyway, this should be relatively easy since ultimately the type is the same.

Good way to test that this is correct is run `CheckTypes` after the transform in the tests. I think that means doing something more sophisticated than just `compile` though.

Reply:
Oh and we also need to update the type of the instance `WRefs`, to be more concrete:
* Inside a module with an overprovisioned port, fix the type of all `WRef(<overprovisioned port name>, ...)`
* In modules that *instantiate* modules with overprovisioned ports fix 3 types:
    * Fix type of the instance itself: `WDefInstance(<instance name>, ...)` (the `BundleType`)
    * Fix type of  `WRef` pointing to the instance (same type as `WDefInstance`)
    * Fix type of `WSubfield` pointing to the port

Reply:
Because this works as is on the *.3 release, do you still think I should update it? Or are we ok as is?